File: shop_staff/plugins/starxpand_flutter/android/src/main/kotlin/com/example/starxpand_flutter/StarxpandFlutterModels.kt
package com.example.starxpand_flutter

import android.util.Log
import com.starmicronics.stario10.InterfaceType
import com.starmicronics.stario10.StarConnectionSettings

internal class StarxpandPluginException(
    val errorCode: String,
    override val message: String,
) : IllegalArgumentException(message)

internal data class StarxpandPrinterTarget(
    val transport: Transport,
    val identifier: String?,
    val host: String?,
    val port: Int?,
    val modelName: String?,
    val autoSwitchInterface: Boolean,
) {
    enum class Transport(val wireValue: String) {
        Network("network"),
        BluetoothClassic("bluetooth_classic"),
        BluetoothLe("bluetooth_le"),
        Usb("usb"),
        UsbC("usb_c"),
        LightningUsb("lightning_usb"),
        ;

        companion object {
            fun fromWireValue(raw: String?): Transport? {
                return entries.firstOrNull { it.wireValue == raw }
            }
        }
    }

    val interfaceType: InterfaceType
        get() = when (transport) {
            Transport.Network -> InterfaceType.Lan
            Transport.BluetoothClassic -> InterfaceType.Bluetooth
            Transport.BluetoothLe -> InterfaceType.BluetoothLE
            Transport.Usb,
            Transport.UsbC,
            -> InterfaceType.Usb
            Transport.LightningUsb -> throw StarxpandPluginException(
                errorCode = "unsupported_transport",
                message = "lightning_usb is only supported on iOS.",
            )
        }

    val resolvedIdentifier: String
        get() = when (transport) {
            Transport.Network -> host ?: identifier ?: ""
            Transport.BluetoothClassic,
            Transport.BluetoothLe,
            Transport.Usb,
            Transport.UsbC,
            Transport.LightningUsb,
            -> identifier ?: host ?: ""
        }

    fun makeConnectionSettings(): StarConnectionSettings {
        val resolved = resolvedIdentifier
        if (resolved.isBlank()) {
            throw StarxpandPluginException(
                errorCode = "invalid_printer",
                message = "Printer identifier is required for ${transport.wireValue}.",
            )
        }

        if (transport == Transport.Network && port != null && port != 9100) {
            Log.d(
                "starxpand_flutter",
                "Ignoring unsupported custom LAN port $port for StarIO10 printer connection.",
            )
        }

        return StarConnectionSettings(interfaceType, resolved, autoSwitchInterface)
    }

    fun requiresBluetoothPermission(): Boolean {
        return transport == Transport.BluetoothClassic || transport == Transport.BluetoothLe
    }

    companion object {
        fun discoveryTransports(json: Map<*, *>): List<Transport> {
            val rawValues = json["transports"] as? List<*> ?: emptyList<Any?>()
            if (rawValues.isEmpty()) {
                return listOf(
                    Transport.Network,
                    Transport.BluetoothClassic,
                    Transport.BluetoothLe,
                    Transport.Usb,
                )
            }

            return rawValues.mapNotNull { raw ->
                Transport.fromWireValue(raw as? String)
            }
        }

        fun fromJson(json: Map<*, *>): StarxpandPrinterTarget {
            val transport = Transport.fromWireValue(json["transport"] as? String)
                ?: throw StarxpandPluginException(
                    errorCode = "invalid_printer",
                    message = "Missing or invalid printer transport.",
                )

            return StarxpandPrinterTarget(
                transport = transport,
                identifier = (json["identifier"] as? String)?.nilIfBlank(),
                host = (json["host"] as? String)?.nilIfBlank(),
                port = intValue(json["port"]),
                modelName = (json["modelName"] as? String)?.nilIfBlank(),
                autoSwitchInterface = boolValue(json["autoSwitchInterface"]) ?: false,
            )
        }
    }
}

internal data class StarxpandDiscoveredPrinter(
    val transport: StarxpandPrinterTarget.Transport,
    val identifier: String,
    val host: String?,
    val modelName: String?,
    val displayName: String?,
    val connectionInfo: String?,
) {
    fun toMap(): Map<String, Any> {
        return buildMap {
            put("transport", transport.wireValue)
            put("identifier", identifier)
            host?.takeIf { it.isNotBlank() }?.let { put("host", it) }
            modelName?.takeIf { it.isNotBlank() }?.let { put("modelName", it) }
            displayName?.takeIf { it.isNotBlank() }?.let { put("displayName", it) }
            connectionInfo?.takeIf { it.isNotBlank() }?.let { put("connectionInfo", it) }
        }
    }
}

internal data class ReceiptPrintPlanDocument(
    val paperWidthMm: Int,
    val nodes: List<ReceiptPrintPlanNode>,
) {
    companion object {
        fun fromJson(json: Map<*, *>): ReceiptPrintPlanDocument {
            val rawNodes = json["nodes"] as? List<*> ?: emptyList<Any?>()
            val nodes = rawNodes.mapNotNull { raw ->
                (raw as? Map<*, *>)?.let(ReceiptPrintPlanNode::fromJson)
            }

            return ReceiptPrintPlanDocument(
                paperWidthMm = intValue(json["paperWidthMm"]) ?: 72,
                nodes = nodes,
            )
        }
    }
}

internal data class ReceiptPrintPlanNode(
    val type: NodeType,
    val text: String?,
    val columns: List<ReceiptPrintPlanColumn>,
    val align: Align,
    val bold: Boolean,
    val widthScale: Int,
    val heightScale: Int,
    val imageAssetKey: String?,
    val imageBase64: String?,
    val qrData: String?,
    val spacerLines: Int?,
    val partialCut: Boolean,
) {
    enum class NodeType(val wireValue: String) {
        Image("image"),
        Text("text"),
        Row("row"),
        Divider("divider"),
        Spacer("spacer"),
        Cut("cut"),
        QrCode("qr_code"),
        ;

        companion object {
            fun fromWireValue(raw: String?): NodeType? {
                return entries.firstOrNull { it.wireValue == raw }
            }
        }
    }

    enum class Align(val wireValue: String) {
        Left("left"),
        Center("center"),
        Right("right"),
        ;

        companion object {
            fun fromWireValue(raw: String?): Align {
                return entries.firstOrNull { it.wireValue == raw } ?: Left
            }
        }
    }

    companion object {
        fun fromJson(json: Map<*, *>): ReceiptPrintPlanNode {
            val type = NodeType.fromWireValue(json["type"] as? String)
                ?: throw StarxpandPluginException(
                    errorCode = "invalid_plan",
                    message = "Unknown plan node type.",
                )

            val rawColumns = json["columns"] as? List<*> ?: emptyList<Any?>()
            val columns = rawColumns.mapNotNull { raw ->
                (raw as? Map<*, *>)?.let(ReceiptPrintPlanColumn::fromJson)
            }

            return ReceiptPrintPlanNode(
                type = type,
                text = (json["text"] as? String)?.nilIfBlank(),
                columns = columns,
                align = Align.fromWireValue(json["align"] as? String),
                bold = boolValue(json["bold"]) ?: false,
                widthScale = intValue(json["widthScale"]) ?: 1,
                heightScale = intValue(json["heightScale"]) ?: 1,
                imageAssetKey = (json["imageAssetKey"] as? String)?.nilIfBlank(),
                imageBase64 = (json["imageBase64"] as? String)?.nilIfBlank(),
                qrData = (json["qrData"] as? String)?.nilIfBlank(),
                spacerLines = intValue(json["spacerLines"]),
                partialCut = boolValue(json["partialCut"]) ?: true,
            )
        }
    }
}

internal data class ReceiptPrintPlanColumn(
    val text: String,
    val align: ReceiptPrintPlanNode.Align,
    val flex: Int,
    val bold: Boolean,
) {
    companion object {
        fun fromJson(json: Map<*, *>): ReceiptPrintPlanColumn {
            return ReceiptPrintPlanColumn(
                text = (json["text"] as? String) ?: "",
                align = ReceiptPrintPlanNode.Align.fromWireValue(json["align"] as? String),
                flex = (intValue(json["flex"]) ?: 1).coerceAtLeast(1),
                bold = boolValue(json["bold"]) ?: false,
            )
        }
    }
}

internal data class ReceiptDocumentContext(
    val locale: String?,
) {
    companion object {
        fun fromJson(json: Map<*, *>): ReceiptDocumentContext {
            val transaction = json["transaction"] as? Map<*, *>
            return ReceiptDocumentContext(
                locale = (transaction?.get("locale") as? String)?.nilIfBlank(),
            )
        }
    }
}

internal fun intValue(value: Any?): Int? {
    return when (value) {
        is Int -> value
        is Long -> value.toInt()
        is Double -> value.toInt()
        is Float -> value.toInt()
        is Number -> value.toInt()
        is String -> value.trim().toIntOrNull()
        else -> null
    }
}

internal fun boolValue(value: Any?): Boolean? {
    return when (value) {
        is Boolean -> value
        is String -> when (value.trim().lowercase()) {
            "true" -> true
            "false" -> false
            else -> null
        }
        else -> null
    }
}

private fun String.nilIfBlank(): String? {
    return if (isBlank()) null else this
}

File: shop_staff/plugins/starxpand_flutter/ios/Classes/StarxpandFlutterModels.swift
import Foundation
import StarIO10

enum StarxpandPluginError: LocalizedError {
  case invalidArguments(String)
  case invalidPrinter(String)
  case invalidPlan(String)
  case missingAsset(String)
  case unsupportedTransport(String)
  case missingPermission(String)
  case printerError(String)

  var errorDescription: String? {
    switch self {
    case .invalidArguments(let message),
        .invalidPrinter(let message),
        .invalidPlan(let message),
        .missingAsset(let message),
        .unsupportedTransport(let message),
        .missingPermission(let message),
        .printerError(let message):
      return message
    }
  }

  var code: String {
    switch self {
    case .invalidArguments:
      return "invalid_arguments"
    case .invalidPrinter:
      return "invalid_printer"
    case .invalidPlan:
      return "invalid_plan"
    case .missingAsset:
      return "missing_asset"
    case .unsupportedTransport:
      return "unsupported_transport"
    case .missingPermission:
      return "missing_permission"
    case .printerError:
      return "printer_error"
    }
  }
}

struct StarxpandPrinterTarget {
  enum Transport: String {
    case network = "network"
    case bluetoothClassic = "bluetooth_classic"
    case bluetoothLe = "bluetooth_le"
    case usb = "usb"
    case usbC = "usb_c"
    case lightningUsb = "lightning_usb"
  }

  static func discoveryTransports(json: [String: Any]) -> [Transport] {
    let rawValues = json["transports"] as? [String] ?? []
    if rawValues.isEmpty {
      return [
        .network,
        .bluetoothClassic,
        .bluetoothLe,
        .lightningUsb
      ]
    }

    return rawValues.compactMap(Transport.init(rawValue:))
  }

  let transport: Transport
  let identifier: String?
  let host: String?
  let port: Int?
  let modelName: String?
  let autoSwitchInterface: Bool

  init(json: [String: Any]) throws {
    guard let rawTransport = json["transport"] as? String,
          let transport = Transport(rawValue: rawTransport)
    else {
      throw StarxpandPluginError.invalidPrinter("Missing or invalid printer transport.")
    }

    self.transport = transport
    self.identifier = (json["identifier"] as? String)?.nilIfEmpty
    self.host = (json["host"] as? String)?.nilIfEmpty
    self.port = json["port"] as? Int
    self.modelName = (json["modelName"] as? String)?.nilIfEmpty
    self.autoSwitchInterface = (json["autoSwitchInterface"] as? Bool) ?? false
  }

  var interfaceType: InterfaceType {
    switch transport {
    case .network:
      return .lan
    case .bluetoothClassic:
      return .bluetooth
    case .bluetoothLe:
      return .bluetoothLE
    case .usb, .usbC, .lightningUsb:
      return .usb
    }
  }

  var resolvedIdentifier: String {
    switch transport {
    case .network:
      return host ?? identifier ?? ""
    case .bluetoothClassic, .bluetoothLe, .usb, .usbC, .lightningUsb:
      return identifier ?? host ?? ""
    }
  }

  func makeConnectionSettings() throws -> StarConnectionSettings {
    let resolved = resolvedIdentifier
    if resolved.isEmpty {
      throw StarxpandPluginError.invalidPrinter(
        "Printer identifier is required for \(transport.rawValue)."
      )
    }

    if transport == .network, let port, port != 9100 {
      // StarIO10 LAN uses the printer identifier only. Keep the host and ignore
      // custom ports for now rather than fabricating an unsupported identifier.
      NSLog("starxpand_flutter: ignoring unsupported custom LAN port \(port)")
    }

    return StarConnectionSettings(
      interfaceType: interfaceType,
      identifier: resolved,
      autoSwitchInterface: autoSwitchInterface
    )
  }
}

struct StarxpandDiscoveredPrinter {
  let transport: StarxpandPrinterTarget.Transport
  let identifier: String
  let host: String?
  let modelName: String?
  let displayName: String?
  let connectionInfo: String?

  var dictionary: [String: Any] {
    var payload: [String: Any] = [
      "transport": transport.rawValue,
      "identifier": identifier
    ]
    if let host, !host.isEmpty {
      payload["host"] = host
    }
    if let modelName, !modelName.isEmpty {
      payload["modelName"] = modelName
    }
    if let displayName, !displayName.isEmpty {
      payload["displayName"] = displayName
    }
    if let connectionInfo, !connectionInfo.isEmpty {
      payload["connectionInfo"] = connectionInfo
    }
    return payload
  }
}

struct StarxpandDrawerStatusPayload {
  let hasError: Bool
  let coverOpen: Bool
  let drawerOpenCloseSignal: Bool
  let paperEmpty: Bool
  let paperNearEmpty: Bool

  init(status: StarPrinterStatus) {
    self.hasError = status.hasError
    self.coverOpen = status.coverOpen
    self.drawerOpenCloseSignal = status.drawerOpenCloseSignal
    self.paperEmpty = status.paperEmpty
    self.paperNearEmpty = status.paperNearEmpty
  }

  var dictionary: [String: Any] {
    [
      "hasError": hasError,
      "coverOpen": coverOpen,
      "drawerOpenCloseSignal": drawerOpenCloseSignal,
      "paperEmpty": paperEmpty,
      "paperNearEmpty": paperNearEmpty
    ]
  }
}

struct StarxpandDrawerOpenRequest {
  enum Channel: String {
    case no1
    case no2

    var starChannel: StarXpandCommand.Drawer.Channel {
      switch self {
      case .no1:
        return .no1
      case .no2:
        return .no2
      }
    }
  }

  let channel: Channel
  let onTimeMs: Int

  init(json: [String: Any]) {
    self.channel = Channel(rawValue: (json["channel"] as? String) ?? "no1") ?? .no1
    self.onTimeMs = max(50, json["onTimeMs"] as? Int ?? 200)
  }
}

struct ReceiptPrintPlanDocument {
  let paperWidthMm: Int
  let nodes: [ReceiptPrintPlanNode]

  init(json: [String: Any]) throws {
    self.paperWidthMm = json["paperWidthMm"] as? Int ?? 72

    let rawNodes = json["nodes"] as? [[String: Any]] ?? []
    self.nodes = try rawNodes.map(ReceiptPrintPlanNode.init(json:))
  }
}

struct ReceiptPrintPlanNode {
  enum NodeType: String {
    case image
    case text
    case row
    case divider
    case spacer
    case cut
    case qrCode = "qr_code"
  }

  enum Align: String {
    case left
    case center
    case right
  }

  let type: NodeType
  let text: String?
  let columns: [ReceiptPrintPlanColumn]
  let align: Align
  let bold: Bool
  let widthScale: Int
  let heightScale: Int
  let imageAssetKey: String?
  let imageBase64: String?
  let qrData: String?
  let spacerLines: Int?
  let partialCut: Bool

  init(json: [String: Any]) throws {
    guard let rawType = json["type"] as? String,
          let type = NodeType(rawValue: rawType)
    else {
      throw StarxpandPluginError.invalidPlan("Unknown plan node type.")
    }

    self.type = type
    self.text = (json["text"] as? String)?.nilIfEmpty
    let rawColumns = json["columns"] as? [[String: Any]] ?? []
    self.columns = rawColumns.map(ReceiptPrintPlanColumn.init(json:))
    self.align = Align(rawValue: (json["align"] as? String) ?? "left") ?? .left
    self.bold = (json["bold"] as? Bool) ?? false
    self.widthScale = json["widthScale"] as? Int ?? 1
    self.heightScale = json["heightScale"] as? Int ?? 1
    self.imageAssetKey = (json["imageAssetKey"] as? String)?.nilIfEmpty
    self.imageBase64 = (json["imageBase64"] as? String)?.nilIfEmpty
    self.qrData = (json["qrData"] as? String)?.nilIfEmpty
    self.spacerLines = json["spacerLines"] as? Int
    self.partialCut = (json["partialCut"] as? Bool) ?? true
  }
}

struct ReceiptPrintPlanColumn {
  let text: String
  let align: ReceiptPrintPlanNode.Align
  let flex: Int
  let bold: Bool

  init(json: [String: Any]) {
    self.text = (json["text"] as? String) ?? ""
    self.align = ReceiptPrintPlanNode.Align(rawValue: (json["align"] as? String) ?? "left") ?? .left
    self.flex = max(1, json["flex"] as? Int ?? 1)
    self.bold = (json["bold"] as? Bool) ?? false
  }
}

struct ReceiptDocumentContext {
  let locale: String?

  init(json: [String: Any]) {
    if let transaction = json["transaction"] as? [String: Any] {
      self.locale = (transaction["locale"] as? String)?.nilIfEmpty
    } else {
      self.locale = nil
    }
  }
}

private extension String {
  var nilIfEmpty: String? {
    isEmpty ? nil : self
  }
}
